### PR TITLE
fix: Upload Timeouts When Batching

### LINKED ISSUE #15: Check JPA improvements in the batch process
- reflect the current implementation
- check possible improvements using more JPA (without reducing legibility)
- optimize the accesses and traffic
- check the use a transaction - high isolation (The initial assumption is to use TRANSACTION_SERIALIZABLE)

SERIALIZABLE is the highest level of isolation. It prevents all mentioned concurrency side effects but can lead to the lowest concurrent access rate because it executes concurrent calls sequentially. In other words, concurrent execution of a group of serializable transactions has the same result as executing them in serial.

### LINKED ISSUE #73: Batching blocks requests, which results in timeouts
## Describe the bug

When batching process is in progress upload requests are blocked until database is available again. This can result in timeouts on client side.

## Expected behaviour

Upload requests should be answered in a few seconds and not timeout.

## Steps to reproduce the issue

1. Upload a big amount of keys (eg. 50 upload requests with 5000 keys each)
2. wait until batching starts
3. requests will timeout

## Possible Fix

Implement Schedule-Lock to limit batch processing to one instance which results in the possibility to remove the transaction level isolation. (--> Keys can be uploaded during batching process)

### PR DESCRIPTION
**Pull request by f11h**
To fix this issue the batching process needed a refactoring.
Now it is much faster and can handle uploads during it is executed.

This PR closes #73 and closes #15

### CODE REVIEW

**Review by mlaue-tech — APPROVED** [review]
Just minor typo to change 😉

File: docs/software-design-federation-gateway-service.md [hunk 1]
```diff
@@ -133,7 +133,7 @@ These key-value-pairs can be followed by additional attributes. The additional a
 | **Batching Service**
 | Batch Process started | INFO | Batch process started | n/a |
 | Batch Process: Created Batch | INFO | Batch created | batchTag, diagnosisKeyCount |
```
Comment (dfischer-tech) [inline docs/software-design-federation-gateway-service.md:135]: This message could documented as well: 
| Batch Process: Successful finish | INFO | successfully finish the document batching process - no more unprocessed diagnosis keys left | n/a |

File: src/main/java/eu/interop/federationgateway/service/DiagnosisKeyBatchService.java [hunk 2]
```diff
@@ -65,158 +61,140 @@
...
+  @Transactional(Transactional.TxType.MANDATORY)
+  public DiagnosisKeyBatchEntity createNextBatchEntityAndLinkPredecessor() {
+    DiagnosisKeyBatchEntity newBatchEntity = new DiagnosisKeyBatchEntity();
+    newBatchEntity.setCreatedAt(ZonedDateTime.now(ZoneOffset.UTC));
+
+    String formattedDate = LocalDate.now().format(DateTimeFormatter.ofPattern("yyyyMMdd"));
+
+    Optional<DiagnosisKeyBatchEntity> lastBatchEntity = diagnosisKeyBatchRepository.findTopByOrderByCreatedAtDesc();
 
-    if (diagnosisKeyEntitys.stream()
```
Comment (dfischer-tech) [inline src/main/java/eu/interop/federationgateway/service/DiagnosisKeyBatchService.java:166]: do not forget the requirement: make sure that within a batch the versions (FormatInformation, majorVersion) of the uploaded data are the same
proposal: separate new versions in separate batches

File: src/main/java/eu/interop/federationgateway/service/DiagnosisKeyBatchService.java [hunk 3]
```diff
@@ -65,158 +61,140 @@
...
-
-    if (lastEntry.isPresent() && isBatchFromCurrentDay(lastEntry, currentDateTime)) {
-      batchEntity.setBatchName(formattedDate + "-" + getSequenceFromPreviousBatchAndCountOn(lastEntry));
-      lastEntry.get().setBatchLink(batchEntity.getBatchName());
-      // update the last batch by link to the new batch
-      diagnosisKeyBatchRepository.save(lastEntry.get());
-      log.info("successfully updated the linked batch entity");
-    } else {
-      batchEntity.setBatchName(formattedDate + "-1");
+      if (newBatchSize + uploadBatchSize <= properties.getBatching().getDoclimit()) {
```
Comment (dfischer-tech) [inline src/main/java/eu/interop/federationgateway/service/DiagnosisKeyBatchService.java:179]: we should discuss about it again, the batch doc limit must always less than or equal to the upload limit
maybe we should take the same parameter

File: src/main/java/eu/interop/federationgateway/service/DiagnosisKeyBatchService.java [hunk 4]
```diff
@@ -65,158 +61,140 @@
...
+  /**
+   * Creates a new Batch Entity and sets batchTag to all contained diagnosiskeys.
+   *
+   * @return true if a batch was created or false if not.
+   */
+  @Transactional(Transactional.TxType.REQUIRES_NEW)
+  public boolean createNextBatch() {
+
+    List<String> uploaderBatchTags = collectUploaderBatchTags();
+    log.info("collected batch tags");
```
Comment (dfischer-tech) [inline src/main/java/eu/interop/federationgateway/service/DiagnosisKeyBatchService.java:101]: possibly remove this message or add values to it and then add it to the documentation

**Review by dfischer-tech — APPROVED** [review]

File: src/main/java/eu/interop/federationgateway/service/DiagnosisKeyBatchService.java (hunk 4 above)
Comment (f11h) [reply src/main/java/eu/interop/federationgateway/service/DiagnosisKeyBatchService.java:101]: Forgot to remove it

**Review by dfischer-tech — APPROVED** [review]
looks fine

**Pull request merged**